errbot/flow.py

```diff
@@ -0,0 +1,335 @@
+from yapsy.IPlugin import IPlugin
+
+from errbot import Message
+from errbot.backends.base import Identifier
+
+log = logging.getLogger(__name__)
+
+Predicate = Callable[[Mapping[str, Any]], str]
+
+
+class FlowNode(object):
+    """
+    This is a step in a Flow/conversation. It is linked to a specific botcmd and also a "predicate".
+
+    The predicate is a function that tells the flow executor if the flow can enter the step without the user
+    intervention (automatically). The predicates defaults to False.
+
+    The predicate is a function that takes one parameter, the context of the conversation.
+    """
+    def __init__(self, command: str=None):
+        """
+        Creates a FlowNone, takes the command to which the Node is linked to.
+        :param command: the command this Node is linked to. Can only be None if this Node is a Root.
+        """
+        self.command = command
+        self.children = []  # (predicate, node)
+
+    def connect(self, node_or_command: Union['FlowNode', str], predicate: Predicate=lambda _: False):
+        """
+        Construct the flow graph by connecting this node to another node or a command.
+        The predicate is a function that tells the flow executor if the flow can enter the step without the user
+        intervention (automatically).
+        :param node_or_command: the node or a string for a command you want to connect this Node to
+                               (this node or command will be the follow up of this one)
+        :param predicate: function with one parameter, the context, to determine of the flow executor can continue
+                           automatically this flow with no user intervention.
+        :return: the newly created node if you passed a command or the node you gave it to be easily chainable.
+        """
+        node_to_connect_to = node_or_command if isinstance(node_or_command, FlowNode) else FlowNode(node_or_command)
+        self.children.append((predicate, node_to_connect_to))
+        return node_to_connect_to
+
+    def predicate_for_node(self, node: 'FlowNode'):
+        """
+        gets the predicate function for the specified child node.
+        :param node: the child node
+        :return: the predicate that allows the automatic execution of that node.
+        """
+        for predicate, possible_node in self.children:
+            if node == possible_node:
+                return predicate
+        return None
+
+    def __str__(self):
+        return self.command
+
+
+class FlowRoot(FlowNode):
+    """
+    This represent the entry point of a flow description.
+    """
+    def __init__(self, name: str, description: str):
+        """
+
+        :param name: The name of the conversation/flow.
+        :param description:  A human description of what this flow does.
+        """
+        super().__init__()
+        self.name = name
+        self.description = description
+        self.auto_triggers = []
```

Comment:
Since auto_triggers seems to only be used for testing membership, would it be better if it was a set instead of a list?